src/features/card/card.slice.ts

```diff
@@ -1,56 +1,144 @@
+      ) => {
+        if (manipulatedCardId !== targetCardId) {
+          // Unexpected card is chosen
+          window.alert("Failed moving the card. Please try again.");
+          return state;
+        }
+      };
+
+      if (startColumnId === endColumnId) {
+        // Reorder cards in the same column.
+        const targetCardList = [...state.cardsByColumnId[startColumnId]];
+        const [targetCard] = targetCardList.splice(startIndex, 1);
+        targetCardList.splice(endIndex, 0, targetCard);
+
+        const updateLinkedListPointers = () => {
+          const nextSourseCardId = targetCardList[startIndex + 1]?.id || null;
+
+          if (startIndex > 0) {
+            // There is a previous card. Need to update next ID of previous card.
+            const prevSourseCard = targetCardList[startIndex - 1];
+            prevSourseCard.nextCardId = nextSourseCardId;
+          } else {
+            // There is no previous card. Need to update head's card ID of target column.
+            // [TODO]: Update headerCardId of column state
+          }
+
+          const nextDestCardId = targetCardList[endIndex + 1]?.id || null;
+          targetCard.nextCardId = nextDestCardId;
+        };
+        updateLinkedListPointers();
+
+        checkCardIds(targetCard.id, targetCardId);
+
+        // [TODO]: Hit APIs to update data in database
+
+        state.cardsByColumnId[startColumnId] = targetCardList;
+        return state;
+      } else {
+        // Reorder cards across different columns.
+        const sourceCardList = [...state.cardsByColumnId[startColumnId]];
+        const [targetCard] = sourceCardList.splice(startIndex, 1);
+        const destCardList = [...state.cardsByColumnId[endColumnId]];
+        destCardList.splice(endIndex, 0, targetCard);
+
+        // [TODO]: Enable to change nextCardId and headCardID then hit API to update data in database.
+
+        checkCardIds(targetCard.id, targetCardId);
+
+        state.cardsByColumnId[startColumnId] = sourceCardList;
+        state.cardsByColumnId[endColumnId] = destCardList;
+      }
     },
   },
   extraReducers: (builder) => {
     builder.addCase(getBoardThunk.fulfilled, (state, action) => {
-      const data = action.payload?.data;
-      if (data) {
-        data.cards
-          ? (state.cards = data.cards.map((card) =>
-              convertCardToInnerType(card)
-            ))
-          : (state.cards = []);
+      const cards = action.payload?.data?.cards;
+      if (!cards) {
+        state.cardsByColumnId = [];
+        return state;
       }
+
+      const cardsByColumnId: Record<number, InnerCard[]> = {};
+      cards
+        .map((card) => convertCardToInnerType(card))
+        .forEach((card) => {
+          if (cardsByColumnId[card.columnId]) {
+            cardsByColumnId[card.columnId].push(card);
+          } else {
+            cardsByColumnId[card.columnId] = [card];
+          }
+        });
+
+      state.cardsByColumnId = cardsByColumnId;
```

Comment:
Store Card's data divided by column to make it easy to manage the order of cards